### ISSUE TITLE
Scan could no longer be monitored but is still running.

### ISSUE THREAD

**Issue by jormungander** [Arachni/arachni#558]
I am running Arachni 1.0.6 with WebUI 0.5.6 on a RedHat EL server with 3.75G of memory.

This is my first time using Arachni, so maybe I missed something simple.

I started running the scan and after an hour or so got the message in the UI "The scan could no longer be monitored because connection with the Instance was lost. The Instance has probably encountered a fatal error and stopped."  Unfortunately, I can't attach the full error log because github says the body is too long, but the relevant portion appears to be at the beginning.

However, I could see in the shell window that the scan was still running.  I eventually killed the process, but when I started it, the scan resumed.

1) How do I run the scan so it doesn't crash?
2) How do I get a running scan to stop?

dmesg says:

```
Out of memory: Kill process 3241 (ruby) score 818 or sacrifice child
Killed process 3241, UID 500, (ruby) total-vm:4481052kB, anon-rss:3043276kB, file-rss:96kB
```

```
OPTIONS:
---
session: {}
http:
  user_agent: Arachni/v1.0.6
  request_timeout: 50000
  request_redirect_limit: 5
  request_concurrency: 20
  request_queue_size: 500
  request_headers: {}
  cookies: {}
scope:
  redundant_path_patterns: {}
  dom_depth_limit: 10
  exclude_path_patterns: []
  exclude_content_patterns: []
  include_path_patterns: []
  restrict_paths: []
  extend_paths: []
  url_rewrites: {}
browser_cluster:
  pool_size: 6
  job_timeout: 120
  worker_time_to_live: 100
  ignore_images: false
  screen_width: 1600
  screen_height: 1200
datastore:
  token: b545cf8b5601156e8dfd96694cffcc93
audit:
  exclude_vector_patterns: []
  include_vector_patterns: []
  link_templates: []
  links: true
  forms: true
  cookies: true
input:
  values:
    "(?i-mx:name)": arachni_name
    "(?i-mx:user)": arachni_user
    "(?i-mx:usr)": arachni_user
    "(?i-mx:pass)": 5543!%arachni_secret
    "(?i-mx:txt)": arachni_text
    "(?i-mx:num)": '132'
    "(?i-mx:amount)": '100'
    "(?i-mx:mail)": arachni@email.gr
    "(?i-mx:account)": '12'
    "(?i-mx:id)": '1'
  default_values:
    "(?i-mx:name)": arachni_name
    "(?i-mx:user)": arachni_user
    "(?i-mx:usr)": arachni_user
    "(?i-mx:pass)": 5543!%arachni_secret
    "(?i-mx:txt)": arachni_text
    "(?i-mx:num)": '132'
    "(?i-mx:amount)": '100'
    "(?i-mx:mail)": arachni@email.gr
    "(?i-mx:account)": '12'
    "(?i-mx:id)": '1'
  without_defaults: true
  force: false
checks:
- allowed_methods
- backdoors
- backup_directories
- backup_files
- captcha
- code_injection
- code_injection_php_input_wrapper
- code_injection_timing
- common_directories
- common_files
- cookie_set_for_parent_domain
- credit_card
- csrf
- cvs_svn_users
- directory_listing
- emails
- file_inclusion
- form_upload
- hsts
- htaccess_limit
- html_objects
- http_only_cookies
- http_put
- insecure_cookies
- interesting_responses
- ldap_injection
- localstart_asp
- mixed_resource
- no_sql_injection
- no_sql_injection_differential
- origin_spoof_access_restriction_bypass
- os_cmd_injection
- os_cmd_injection_timing
- password_autocomplete
- path_traversal
- private_ip
- response_splitting
- rfi
- session_fixation
- source_code_disclosure
- sql_injection
- sql_injection_differential
- sql_injection_timing
- ssn
- trainer
- unencrypted_password_forms
- unvalidated_redirect
- webdav
- xpath_injection
- xss
- xss_dom
- xss_dom_inputs
- xss_dom_script_context
- xss_event
- xss_path
- xss_script_context
- xss_tag
- xst
platforms: []
plugins:
  autothrottle: {}
  discovery: {}
  healthmap: {}
  timing_attacks: {}
  uniformity: {}
no_fingerprinting: false
authorized_by: 

--------------------------------------------------------------------------------
[2015-03-06 20:34:17 -0500] [RegexpError] failed to allocate memory: /(root|mail):.+:\d+:\d+:.+:[0-9a-zA-Z\/]+/mi
[2015-03-06 20:34:17 -0500] /home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-1.0.6/lib/arachni/element/capabilities/analyzable/taint.rb:164:in `scan'
```

**Comment by Zapotek** [comment]
Hello there,

Depending on the website, memory consumption could rise but there are ways to optimize the scan to match your resources: http://support.arachni-scanner.com/kb/general-use/optimizing-for-faster-scans

So you can either follow these instructions to limit memory consumption or use a machine with more RAM.

However, the site you're scanning doesn't seem to have any of the usual attributes that would lead to a lot of RAM usage so, if you'll allow it, I'd like to perform an identical scan in order to make sure that this isn't due to a memory leak.

As for the scan still running, are we talking about the same scan? Because that seems improbable as it was killed by the kernel.
What made you think that the scan was still running?

Regarding stopping scans, that needs to happen via the WebUI, if you start killing processes then the situation becomes a bit of a hustle.
Although if we're still talking about the improbably runaway scan with which the WebUI has lost communication, then you've got no option other than to kill its process and any browser processes it may have started.

If you just want to kill everything then you need to kill them by their Group ID and not their PID, as the interface, the scanners it'll have spawned and any browsers the scanners will have spawned will belong to the same group.

So, to sum up:

1. May I please run an identical scan to look into the RAM consumption situation?
2. What makes you think that the scan that the kernel reports it has killed was still running? Or are those 2 separate occurrences?

Cheers

**Comment by jormungander** [comment]
1) Sorry, I can't give you permission to run a scan, it's against Amazon's TOS where the server is running.  All penetration tests have to be pre-approved and it takes several days.

Is there something I could do locally that would give you more information?

2) The window where I run arachni_web looked like it was printing things out as the scan ran, and I thought it was relevant to the scan, but what it's printing out when running now is just the same thing over and over again.

I tried running it again, and saw this error after a about 7 minutes:
#<ActiveRecord::StatementInvalid:0x07cdcd08
    attr_reader :original_exception = #<SQLite3::BusyException: database is locked>
>
[
    [ 0] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:108:in `step'",
    [ 1] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:108:in `block in each'",
    [ 2] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:107:in `loop'",
    [ 3] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/statement.rb:107:in `each'",
    [ 4] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:149:in `map'",
    [ 5] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:149:in `block in execute'",
    [ 6] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:95:in `prepare'",
    [ 7] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:134:in `execute'",
    [ 8] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/sqlite3-1.3.10/lib/sqlite3/database.rb:517:in `commit'",
    [ 9] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/sqlite3_adapter.rb:362:in `block in commit_db_transaction'",
    [10] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract_adapter.rb:373:in `block in log'",
    [11] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activesupport-4.1.8/lib/active_support/notifications/instrumenter.rb:20:in `instrument'",
    [12] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract_adapter.rb:367:in `log'",
    [13] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/sqlite3_adapter.rb:362:in `commit_db_transaction'",
    [14] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/transaction.rb:180:in `perform_commit'",
    [15] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/transaction.rb:120:in `commit'",
    [16] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:235:in `commit_transaction'",
    [17] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:215:in `within_new_transaction'",
    [18] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/connection_adapters/abstract/database_statements.rb:201:in `transaction'",
    [19] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/transactions.rb:208:in `transaction'",
    [20] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/transactions.rb:326:in `with_transaction_returning_status'",
    [21] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/transactions.rb:268:in `block in save'",
    [22] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/transactions.rb:283:in `rollback_active_record_state!'",
    [23] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/activerecord-4.1.8/lib/active_record/transactions.rb:267:in `save'",
    [24] "/home/ec2-user/arachni-1.0.6-0.5.6/system/arachni-ui-web/app/models/scan.rb:541:in `block in refresh'",
    [25] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-rpc-0.2.1.1/lib/arachni/rpc/proxy.rb:58:in `call'",
    [26] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-rpc-0.2.1.1/lib/arachni/rpc/proxy.rb:58:in `block (2 levels) in translate'",
    [27] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-rpc-0.2.1.1/lib/arachni/rpc/client/handler.rb:77:in `call'",
    [28] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-rpc-0.2.1.1/lib/arachni/rpc/client/handler.rb:77:in `receive_response'",
    [29] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-rpc-0.2.1.1/lib/arachni/rpc/client/handler.rb:137:in `receive_object'",
    [30] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-rpc-0.2.1.1/lib/arachni/rpc/protocol.rb:52:in `on_read'",
    [31] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor/connection.rb:236:in `block in _read'",
    [32] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor/connection/error.rb:26:in `call'",
    [33] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor/connection/error.rb:26:in `translate'",
    [34] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor/connection.rb:235:in `_read'",
    [35] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:568:in `each'",
    [36] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:568:in `block in process_connections'",
    [37] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:568:in `each'",
    [38] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:568:in `process_connections'",
    [39] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:317:in `block in run'",
    [40] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:308:in `loop'",
    [41] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:308:in `run'",
    [42] "/home/ec2-user/arachni-1.0.6-0.5.6/system/gems/gems/arachni-reactor-0.1.0.beta5/lib/arachni/reactor.rb:350:in `block in run_in_thread'"

**Comment by Zapotek** [comment]
1. Unfortunately not, I'll need to closely monitor what's going on during the scan in order to determine if there's a memory/object leak.
2. That's solvable. As the welcome screen says, the bundled SQLite3 DB isn't really suitable for larger workloads, you need to [setup PostgreSQL](https://github.com/Arachni/arachni-ui-web/wiki/database#PostgreSQL) in order to use it seriously.

So, I'm afraid I'll have to close this for now.

**Comment by jormungander** [comment]
OK, I'll take a look at the optimizing link and see if I can get it to finish with some of the ideas there.
